plugins/inputs/t128_graphql/t128_graphql.go

```diff
@@ -0,0 +1,284 @@
+#   vlan = "vlan"
+#   ipAddress = "ipAddress"
+#   destinationMac = "destinationMac"
+`
+
+func (*T128GraphQL) SampleConfig() string {
+	return sampleConfig
+}
+
+func (*T128GraphQL) Description() string {
+	return "Read data from a graphQL query"
+}
+
+func (plugin *T128GraphQL) Init() error {
+	if plugin.CollectorName == "" {
+		return fmt.Errorf("collector_name is a required configuration field")
+	}
+
+	if plugin.BaseURL == "" {
+		return fmt.Errorf("base_url is a required configuration field")
+	}
+
+	if plugin.BaseURL[len(plugin.BaseURL)-1:] != "/" {
+		plugin.BaseURL += "/"
+	}
+
+	transport := http.DefaultTransport
+
+	plugin.Client = &http.Client{Transport: transport, Timeout: DefaultRequestTimeout}
+	plugin.Query = plugin.buildQuery()
+
+	return nil
+}
+
+func (plugin *T128GraphQL) Gather(acc telegraf.Accumulator) error {
+	timestamp := time.Now().Round(time.Second)
+
+	plugin.retrieveMetrics(acc, timestamp)
+
+	return nil
+}
+
+func (plugin *T128GraphQL) buildQuery() string {
+	//allow multiple inputs like names[ComboEast,ComboWest] ?
+	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
+	query := "query MyQuery{" + replacer.Replace(plugin.EntryPoint) + "{"
+
+	for _, element := range plugin.Fields {
+		query += "\n" + element
+	}
+	query = strings.TrimSpace(query)
+	for _, element := range plugin.Tags {
+		query += "\n" + element
+	}
+
+	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+
+	return query
+}
+
+func (plugin *T128GraphQL) retrieveMetrics(acc telegraf.Accumulator, timestamp time.Time) {
+	request, err := plugin.createRequest()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to create a request for query %s: %w", plugin.Query, err))
+		return
+	}
+
+	response, err := plugin.Client.Do(request)
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(response.Body)
+
+	if err != nil {
+		template := "query error for metric %s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || response.StatusCode >= 300 {
+		template := fmt.Sprintf("status code %d not OK for metric ", response.StatusCode) + "%s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+
+	//decode json
+	jsonParsed, err := gabs.ParseJSON(buf.Bytes())
+	if err != nil {
+		acc.AddError(fmt.Errorf("invalid json response for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	jsonEntryPoint := plugin.BuildJsonEntryPoint()
+	jsonObj, err := jsonParsed.JSONPointer(jsonEntryPoint)
+	if err != nil {
+		template := "unexpected response for metric %s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+
+	//update acc
+	jsonChildren, err := jsonObj.Children()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to iterate on response nodes for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	//TODO: allow nested fields/tags
+	for _, child := range jsonChildren {
```

Comment:
imo all this should also be a seperate function or typed strcut like (`queryResponseParser.Parse`) that can be individually unit tested as well. its pretty complicated and i think youll need to cover more edges in your units

Reply:
- [ ] separate function for response processing
- [ ] more edges in response processing tests